### ISSUE TITLE
Shadow doesn't recurse parent directories for node_modules

### ISSUE THREAD

**Issue by glittershark** [thheller/shadow-cljs#489]
Specifically this is a case of [yarn workspaces](https://yarnpkg.com/en/docs/workspaces/) being used for a monorepo - we have a directory structure that looks like this:

```
project_root/
├── .git/
├── package.json
├── yarn.lock
├── node_modules/
├── lib-1/
│   ├── package.json
│   └── src/
├── app-1/ (cljs app)
│   ├── package.json (with deps on lib-1)
│   ├── shadow-cljs.edn
│   └── src/
└── app-2/ (node app)
    ├── package.json (with deps on lib-1)
    └── src/
```

Note that `node_modules` is in the project root, not any of the individual apps - this is the way yarn lays out workspace installs.

node apps (`app-2`, in the above diagram) properly recurse up to parent directories to find `node_modules`, but shadow-cljs (`app-1`) complains that they're missing.

**Comment by thheller** [comment]
It is by design that `shadow-cljs` doesn't leave the project directory looking for dependencies.

You can however set `:js-options {:node-modules-dir "../"}` to direct it to use the parent directory when looking for `node_modules`.

**Comment by glittershark** [comment]
Wouldn't it be better to just do what node does here?

**Comment by glittershark** [comment]
Also that option doesn't appear to be having an effect:
```
❯ grep node-modules-dir shadow-cljs.edn
 :js-options   {:node-modules-dir "../"}

❯ shadow-cljs compile app
[:app] Compiling ...
(src/main)
Watching src/main for changes
The required JS dependency "react" is not available, it was required by "cljsjs/react.cljs".

Searched in:<PROJECT-ROOT>/app-1/node_modules

You probably need to run:
  npm install react

See: https://shadow-cljs.github.io/docs/UsersGuide.html#npm-install



❯ test -d ../node_modules/react; echo $?
0
```

**Comment by glittershark** [comment]
oops, had to add it to the `build`. Getting this after doing so:
```
ExceptionInfo: files outside the project are not allowed: <PROJECT-ROOT>/node_modules/react/index.js
```

**Comment by thheller** [comment]
> Wouldn't it be better to just do what node does here?

I don't think so no.

As far I can tell nothing is stopping you from doing
```
cd app-1
npm install
```

which would suddenly leave you with a `app-1/node_modules` dir which would "break" the resolution you actually want since there is now ambiguity which folder should be used for dependencies. Maybe `yarn` actually stops you from running `yarn install` if it detects a parent workspaces config. I don't actually know.

`shadow-cljs` also doesn't support "nested" installs as a way to deal with version conflicts (eg. `node_modules/some-lib/node_modules/some-dep`) since I can't of a single use case where that is ever ok in a browser-targetted build (ie. including the `some-dep` depdendency twice).

> files outside the project are not allowed

Appears that that check might be a bit too strict since an explicit path should be supported.

**Comment by glittershark** [comment]
the use case here is development on `lib-1` during development on `app-1` - intra-workspace dependencies are symlinked by yarn during a workspace install rather than copied like they normally would be, meaning we don't have to run `yarn install` every time we make changes to the shared lib

**Comment by glittershark** [comment]
my current workaround is to just symlink everything into the `node_modules` of the individual projects in a yarn postinstall script, but that's really just a hack to make shadow work

**Comment by thheller** [comment]
The "files outside the project" issue should be fixed in `shadow-cljs@2.8.35`.

**Comment by glittershark** [comment]
cool, thanks for that. Maybe we wanna add something to the user's guide about the config for other users of yarn workspaces?
